### ISSUE TITLE
Fine tune model when the same sentence has multiple positives/answers

### ISSUE THREAD

**Issue by Jakobhenningjensen** [UKPLab/sentence-transformers#2423]
Say I have a dataset consisting of `s1` and `s2` where `s1` and `s2` always "have something in common" e.g `s1`  being the banktransaction text and `s2` being the text on the reciept, extracted by some software.
I then want to fine-tune a bi-/cross encoder (prefer a bi-encoder) such that I for a given banktransaction can get the most similar/probable recipt

```
s1                                                         |       s2
-----------------------------------------+------------------------------------
purchase at wallmart                               <recipt text for wallmart 2020-01-01>
purchase at wallmart                               <recipt text for wallmart 2020-01-03>
purchase at wallmart                               <recipt text for wallmart 2020-01-06>
VDK: Netflix                                       <recipt text for netflix 2020-01-30>
VDK: Netflix                                       <recipt text for netflix 2020-02-28>
```
I'm struggling a bit to find the correct loss-function.
My first intuition would be the `SoftMax`  with one label for each `s1` e.g `purchase at wallmart` group would have `label=1` and `VDK: Netflix` group would have `label=2`. Is that doable in "one go" i.e can I do something like
```
[InputExample(texts = ["purchase at wallmart",   <s2_0>, <s2_1>, <s2_2>], label = 1), 
InputExample(texts = ["VDK: Netflix",   <s2_3>, <s2_4>], label = 2)
```
or does it have to be like the [softmax example](https://www.sbert.net/docs/package_reference/losses.html#softmaxloss) e.g

```
[InputExample(texts = ["purchase at wallmart, <s2_0>",  purchase at wallmart, <s2_1>"], label = 1), 
[InputExample(texts = ["purchase at wallmart, <s2_1>",  purchase at wallmart, <s2_2>"], label = 1), 
[InputExample(texts = ["purchase at wallmart, <s2_2>",  purchase at wallmart, <s2_0>"], label = 1), 
InputExample(texts = ["VDK: Netflix, <s2_3>", "VDK: Netflix, <s2_4>"], label = 2),
```
and what do we do if there is a comma in the sentence (I assume we need to get rid of those)?

Alternatively it could be the `MegaBatchMarginLoss` again with `(s1_i, s2_i)` pairs but in my head that might causes some issues, since I we have the exact same text e.g `"purchase at wallmart"` which might get `s2_0` as the positive and later on get `s2_0` as the negative, which I don't think is very helpfull.

Is there a way to group identical "questions" in a preferable manner e.g `{ "purchase at wallmart": [s2_0, s2_1, s2_2], "VDK: Netflix : [s2_3, s2_4]}` or how do we handle this "same question - different answers"?

**Comment by tomaarsen** [comment]
Hello!

If we want to train a bi-encoder, i.e. a `SentenceTransformer` model in our case, then we can use one of [these losses](https://www.sbert.net/docs/package_reference/losses.html), but it seems you've already figured that out.
In hindsight, the Softmax example is a bit unintuitive: neither of your examples would work. To use the SoftmaxLoss, you should use:
```python
[
InputExample(texts = ["purchase at wallmart", "<s2_0>"], label = 1),
InputExample(texts = ["purchase at wallmart", "<s2_1>"], label = 1),
InputExample(texts = ["purchase at wallmart", "<s2_2>"], label = 1),
InputExample(texts = ["VDK: Netflix", "<s2_3>"], label = 1),
InputExample(texts = ["VDK: Netflix", "<s2_4>"], label = 1),
]
```
Where the label is between 0 and 1.

However, a more popular loss is [`MultipleNegativesRankingLoss`](https://www.sbert.net/docs/package_reference/losses.html#multiplenegativesrankingloss). See also [this documentation](https://www.sbert.net/examples/training/nli/README.html) for more information on it and its relation to the SoftmaxLoss. As you can read in the documentation, it will take positive pairs in a batch `(a_1, p_1), (a_2, p_2)…, (a_n, p_n)`, and then consider `a_i` and `p_i` as a positive pair & `a_i` and `p_j` with `j != i` as a bunch of negative pairs. So, with a batch size of 32, you would have 32 positive pairs and 32 * 31 = 992 negative pairs. This produces great embedding models, but it has a catch in your scenario: If you have a lot of situations where a sentence is similar to a bunch of other sentences, then it's possible that e.g. `a_1 == a_8` in a certain batch during training. Then you will get that `a_1` and `p_1` and `a_8` and `p_8` as positive pairs, but also `a_1` and `p_8` & `a_8` and `p_1` as negative pairs. Because `a_1 == a_8`, the same pairs are both positive & negative here, which may prevent the model from effectively learning these samples. 

Now, if you have a lot of data, and each sentence is only similar to a few other sentences, then perhaps it's very rare that two of the same sentences end up in a batch, so then it wouldn't matter much. This kind of depends on your situation.

---

An alternative, commonly used loss is [`CosineSimilarityLoss`](https://www.sbert.net/docs/package_reference/losses.html#cosinesimilarityloss). It's described [here](https://www.sbert.net/examples/training/sts/README.html) for training Semantic Textual Similarity models. 

As you may have noticed, Sentence Transformers also implements ~11 other loss functions, but in truth, I have yet to see anyone use them.

I would do the following:
* Figure out if you could use MultipleNegativesRankingLoss, i.e., figure out if you would often have batches with the same sentences in them that might inhibit the training.
* If you can't use MNRL, use CosineSimilarityLoss. 

However, feel free to experiment with other losses, and you may even combine losses like so: https://www.sbert.net/examples/training/quora_duplicate_questions/README.html#multi-task-learning. There simply isn't a definitive answer on the best approach.

- Tom Aarsen

**Comment by Jakobhenningjensen** [comment]
First, thanks for the swift reply!

>In hindsight, the Softmax example is a bit unintuitive: neither of your examples would work. To use the SoftmaxLoss, you should use:
>
>[
>InputExample(texts = ["purchase at wallmart", "<s2_0>"], label = 1),
>InputExample(texts = ["purchase at wallmart", "<s2_1>"], label = 1),
>InputExample(texts = ["purchase at wallmart", "<s2_2>"], label = 1),
>InputExample(texts = ["VDK: Netflix", "<s2_3>"], label = 1),
>InputExample(texts = ["VDK: Netflix", "<s2_4>"], label = 1),
>]
>Where the label is between 0 and 1.

is there a typo in the docs/example for `SoftMax` then respect to the labels and the texts-structure? 


 >Then you will get that a_1 and p_1 and a_8 and p_8 as positive pairs, but also a_1 and p_8 & a_8 and p_1 as negative pairs. Because a_1 == a_8, the same pairs are both positive & negative here, which may prevent the model from effectively learning these samples.

exactly, that is my concern aswell, but I'll give it a shot!
--------------

Just a question; with all these hard-labels i.e  positive/negative pairs, why don't we just end up creating a lot of cluster-pairs in each "corned" of the vector-space? Say we have the pairs:

```
q1= ("what is the capital of germany?", "the capital of germany is Berlin")
q2 = ("Has the capital of Germany always been Berlin?", "Germany have had 3 captials; Hamburg, Flensburg and Berlin")
q3 = ("What is a dog?", "A dog is an animal")
```

Now, how does it happen that (hopefully) `cos_sim(q1[0], q2[1])> cos_sim(q1[0], q3[1])`  i.e how does having hard-labels create embeddings that makes `q1` closer to `q2` than `q1` is to `q3`  when both of the answers from `q2` and `q3` would be negatives in the [MultipleNegativesRankingLoss](https://www.sbert.net/docs/package_reference/losses.html#multiplenegativesrankingloss)  setting?

**Comment by tomaarsen** [comment]
> is there a typo in the docs/example for SoftMax then respect to the labels and the texts-structure?

I don't believe so, but it's a bit confusing. Presumably you are referring to this bit:
![image](https://github.com/UKPLab/sentence-transformers/assets/37621491/cbe0d155-4607-4ca2-95c5-1c68b327cf0f)
Some of the commas are inside of the strings. That's what you're referring to right?

---

As for your last question: Presumably through sufficient pairs & training it eventually learns. Indeed, to my knowledge, both q2[1] and q3[1] will be considered negative pairs & their vectors will be pushed apart equivalently. A common strategy here is to adopt multiple losses. As mentioned in the [docs here](https://www.sbert.net/examples/training/quora_duplicate_questions/README.html#multi-task-learning):
> MultipleNegativesRankingLoss on the other sides mainly reduces the distance between positive pairs out of large set of possible candidates.

They use `OnlineContrastiveLoss` here as a second loss, as it should help push negative pairs further away. To my knowledge, CosineSimilarityLoss + MNRL is a fairly common modern combination.

But I think sometimes MNRL is adopted on its own, and it tends to work well. There's a few scripts with it in the [examples](https://github.com/UKPLab/sentence-transformers/tree/master/examples).

- Tom Aarsen

**Comment by tomaarsen** [comment]
Feel free to ping back with your findings! I'll be very curious to read what you've encountered :)

**Comment by Jakobhenningjensen** [comment]
Just one quick thought; Is it possible to "overfit" a sentence transformer in some way when we have a fair amount of same/very alike sentences?

**Comment by tomaarsen** [comment]
It's possible for overfit a ST model, yes. That said, I'm not sure if it's more likely to happen when you have a lot of similar/same sentences. I would use an [evaluator](https://sbert.net/docs/package_reference/evaluation.html) to track the performance on a hold-out set to get a feel of the "real" performance. Ideally, this hold-out set matches roughly what you would feed the model in production.
